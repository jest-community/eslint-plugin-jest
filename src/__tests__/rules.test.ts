import { existsSync } from 'fs';
import { resolve } from 'path';
import plugin from '../';

const ruleNames = Object.keys(plugin.rules);
const numberOfRules = 42;

describe('rules', () => {
  it('should have a corresponding doc for each rule', () => {
    ruleNames.forEach(rule => {
      const docPath = resolve(__dirname, '../../docs/rules', `${rule}.md`);

      if (!existsSync(docPath)) {
        throw new Error(
          `Could not find documentation file for rule "${rule}" in path "${docPath}"`,
        );
      }
    });
  });

  it('should have a corresponding test for each rule', () => {
    ruleNames.forEach(rule => {
      const testPath = resolve(
        __dirname,
        '../rules/__tests__/',
        `${rule}.test.ts`,
      );

      if (!existsSync(testPath)) {
        throw new Error(
          `Could not find test file for rule "${rule}" in path "${testPath}"`,
        );
      }
    });
  });

  it('should have the correct amount of rules', () => {
    const { length } = ruleNames;
    if (length !== numberOfRules) {
      throw new Error(
        `There should be exactly ${numberOfRules} rules, but there are ${length}. If you've added a new rule, please update this number.`,
      );
    }
  });

  it('should export configs that refer to actual rules', () => {
    const recommendedConfigs = plugin.configs;

    expect(recommendedConfigs).toMatchSnapshot();
    expect(Object.keys(recommendedConfigs)).toEqual([
      'all',
      'recommended',
      'style',
    ]);
    expect(Object.keys(recommendedConfigs.all.rules)).toHaveLength(
      ruleNames.length,
    );
    const allConfigRules = Object.values(recommendedConfigs)
      .map(config => Object.keys(config.rules))
      .reduce((previousValue, currentValue) => [
        ...previousValue,
        ...currentValue,
      ]);

    allConfigRules.forEach(rule => {
      const ruleNamePrefix = 'jest/';
      const ruleName = rule.slice(ruleNamePrefix.length);
      expect(rule.startsWith(ruleNamePrefix)).toBe(true);
      expect(ruleNames).toContain(ruleName);
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      expect(() => require(`../rules/${ruleName}`)).not.toThrow();
    });
  });
});
