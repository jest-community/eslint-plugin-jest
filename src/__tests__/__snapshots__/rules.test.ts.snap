// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rules should export configs that refer to actual rules 1`] = `
Object {
  "all": Object {
    "env": Object {
      "jest/globals": true,
    },
    "plugins": Array [
      "jest",
    ],
    "rules": Object {
      "jest/consistent-test-it": "error",
      "jest/expect-expect": "error",
      "jest/lowercase-name": "error",
      "jest/no-alias-methods": "error",
      "jest/no-commented-out-tests": "error",
      "jest/no-disabled-tests": "error",
      "jest/no-duplicate-hooks": "error",
      "jest/no-expect-resolves": "error",
      "jest/no-export": "error",
      "jest/no-focused-tests": "error",
      "jest/no-hooks": "error",
      "jest/no-identical-title": "error",
      "jest/no-if": "error",
      "jest/no-jasmine-globals": "error",
      "jest/no-jest-import": "error",
      "jest/no-large-snapshots": "error",
      "jest/no-mocks-import": "error",
      "jest/no-standalone-expect": "error",
      "jest/no-test-callback": "error",
      "jest/no-test-prefixes": "error",
      "jest/no-test-return-statement": "error",
      "jest/no-truthy-falsy": "error",
      "jest/no-try-expect": "error",
      "jest/prefer-called-with": "error",
      "jest/prefer-expect-assertions": "error",
      "jest/prefer-hooks-on-top": "error",
      "jest/prefer-inline-snapshots": "error",
      "jest/prefer-spy-on": "error",
      "jest/prefer-strict-equal": "error",
      "jest/prefer-to-be-null": "error",
      "jest/prefer-to-be-undefined": "error",
      "jest/prefer-to-contain": "error",
      "jest/prefer-to-have-length": "error",
      "jest/prefer-todo": "error",
      "jest/require-to-throw-message": "error",
      "jest/require-top-level-describe": "error",
      "jest/valid-describe": "error",
      "jest/valid-expect": "error",
      "jest/valid-expect-in-promise": "error",
      "jest/valid-title": "error",
    },
  },
  "recommended": Object {
    "env": Object {
      "jest/globals": true,
    },
    "plugins": Array [
      "jest",
    ],
    "rules": Object {
      "jest/expect-expect": "warn",
      "jest/no-commented-out-tests": "warn",
      "jest/no-disabled-tests": "warn",
      "jest/no-export": "error",
      "jest/no-focused-tests": "error",
      "jest/no-identical-title": "error",
      "jest/no-jasmine-globals": "warn",
      "jest/no-jest-import": "error",
      "jest/no-mocks-import": "error",
      "jest/no-standalone-expect": "error",
      "jest/no-test-callback": "error",
      "jest/no-test-prefixes": "error",
      "jest/no-try-expect": "error",
      "jest/valid-describe": "error",
      "jest/valid-expect": "error",
      "jest/valid-expect-in-promise": "error",
    },
  },
  "style": Object {
    "plugins": Array [
      "jest",
    ],
    "rules": Object {
      "jest/no-alias-methods": "warn",
      "jest/prefer-to-be-null": "error",
      "jest/prefer-to-be-undefined": "error",
      "jest/prefer-to-contain": "error",
      "jest/prefer-to-have-length": "error",
    },
  },
}
`;
